File: ios/HybridCreationTask.swift
import Foundation
import NitroModules

final class HybridCreationTask: HybridNativeCreationTaskSpec {
  private let stateMachine = TaskStateMachine()
  private let operationId: String
  private let engine: ArchiveEngine
  private let request: NativeCreationRequest
  private let useCase: CreateArchiveUseCase
  // ponytail: copy buffers on JS thread before background hop (JSArrayBuffer thread check)
  private let copiedData: [Data?]

  private var progressCallback: ((NativeProgress) -> Void)?

  private var cachedState: String = "idle"
  private var cachedProgress: NativeProgress

  var state: String {
    cachedState
  }

  var progress: NativeProgress {
    cachedProgress
  }

  init(engine: ArchiveEngine, request: NativeCreationRequest) {
    self.operationId = UUID().uuidString
    self.engine = engine
    self.request = request
    self.useCase = CreateArchiveUseCase(engine: engine)
    // ponytail: copy ArrayBuffer data while still on JS thread
    self.copiedData = request.entries.map { entry in
      entry.data.map { Data(bytes: $0.data, count: $0.size) }
    }
    self.cachedProgress = NativeProgress(
      operationId: "",
      phase: "preparing",
      processedBytes: 0,
      totalBytes: nil,
      processedEntries: 0,
      totalEntries: nil,
      currentEntry: nil,
      bytesPerSecond: nil,
      estimatedSecondsRemaining: nil,
      percentage: nil
    )
    super.init()
  }

  func start() throws -> Promise<NativeCreationResult> {
    Promise.async {
      try self.stateMachine.start()
      self.cachedState = "running"
      let startTime = CFAbsoluteTimeGetCurrent()

      do {
        self.emitProgress(phase: "preparing")

        guard self.request.destinationKind == "file", let destinationPath = self.request.destinationPath else {
          throw ArchiveDomainError.providerUnsupported
        }

        if self.request.existingDestination == "error",
           FileManager.default.fileExists(atPath: destinationPath) {
          throw ArchiveDomainError.destinationExists(destinationPath)
        }

        let entryInputs = self.request.entries.enumerated().map { idx, nativeEntry -> EntryInput in
          EntryInput(
            kind: nativeEntry.kind,
            sourcePath: nativeEntry.sourcePath,
            archivePath: nativeEntry.archivePath,
            data: self.copiedData[idx],
            recursive: nativeEntry.recursive,
            includeHidden: nativeEntry.includeHidden,
            followSymlinks: nativeEntry.followSymlinks,
            compressionMethod: nativeEntry.compressionMethod,
            modifiedAt: nativeEntry.modifiedAt
          )
        }

        let plan = CreationPlan(
          entries: entryInputs,
          compressionProfile: self.request.compressionProfile,
          compressionLevel: self.request.compressionLevel.map(Int.init),
          storeAlreadyCompressed: self.request.storeAlreadyCompressed ?? true,
          encryptionMethod: self.request.encryptionMethod,
          encryptionPassword: self.request.encryptionPassword
        )

        let output = LocalFileOutput(path: destinationPath)

        self.emitProgress(phase: "compressing")

        let result = try await self.useCase.execute(
          plan: plan,
          output: output,
          onProgress: { snapshot in
            self.emitProgress(
              phase: "compressing",
              processedBytes: snapshot.processedBytes,
              processedEntries: snapshot.processedEntries,
              currentEntry: snapshot.currentEntry,
              totalEntries: snapshot.totalEntries
            )
          }
        )

        self.emitProgress(phase: "finalizing")
        try self.stateMachine.succeed()
        self.cachedState = "succeeded"

        let duration = (CFAbsoluteTimeGetCurrent() - startTime) * 1000

        self.emitProgress(
          phase: "finalizing",
          processedBytes: result.inputBytes,
          processedEntries: result.entryCount,
          currentEntry: nil,
          totalEntries: result.entryCount,
          percentage: 100
        )

        return NativeCreationResult(
          operationId: self.operationId,
          entryCount: Double(result.entryCount),
          inputBytes: result.inputBytes,
          outputBytes: result.outputBytes,
          durationMs: duration,
          atomicWriteApplied: false,
          warnings: []
        )
      } catch {
        self.stateMachine.fail()
        self.cachedState = "failed"
        throw error
      }
    }
  }

  func cancel() throws -> Bool {
    let didCancel = stateMachine.cancel()
    if didCancel {
      cachedState = "cancelled"
    }
    return didCancel
  }

  func onProgress(callback: @escaping (NativeProgress) -> Void) throws -> () -> Void {
    progressCallback = callback
    return { [weak self] in
      self?.progressCallback = nil
    }
  }

  private func emitProgress(
    phase: String,
    processedBytes: UInt64 = 0,
    processedEntries: Int = 0,
    currentEntry: String? = nil,
    totalEntries: Int? = nil,
    percentage: Double? = nil
  ) {
    let total = totalEntries.map(Double.init)
    let pct: Double?
    if let p = percentage {
      pct = p
    } else if let t = total, t > 0 {
      pct = min(Double(processedEntries) / t * 100, 100)
    } else {
      pct = nil
    }

    let snapshot = NativeProgress(
      operationId: operationId,
      phase: phase,
      processedBytes: processedBytes,
      totalBytes: nil,
      processedEntries: Double(processedEntries),
      totalEntries: total,
      currentEntry: currentEntry,
      bytesPerSecond: nil,
      estimatedSecondsRemaining: nil,
      percentage: pct
    )

    cachedProgress = snapshot
    progressCallback?(snapshot)
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/HybridCreationTask.kt
package com.mcodex.nitroarchive

import com.margelo.nitro.archive.*
import com.margelo.nitro.core.Promise
import com.mcodex.nitroarchive.application.CreateArchiveUseCase
import com.mcodex.nitroarchive.application.ports.*
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.domain.TaskStateMachine
import com.mcodex.nitroarchive.infrastructure.storage.LocalFileOutput
import java.util.UUID

class HybridCreationTask(
    private val engine: ArchiveEngine,
    private val request: NativeCreationRequest
) : HybridNativeCreationTaskSpec() {

    private val stateMachine = TaskStateMachine()
    private val operationId = UUID.randomUUID().toString()
    private val listeners = mutableListOf<(NativeProgress) -> Unit>()
    private var latestProgress = NativeProgress(
        operationId = operationId,
        phase = "idle",
        processedBytes = 0L,
        totalBytes = null,
        processedEntries = 0.0,
        totalEntries = null,
        currentEntry = null,
        bytesPerSecond = null,
        estimatedSecondsRemaining = null,
        percentage = null
    )

    override val state: String get() = stateMachine.state.name.lowercase()
    override val progress: NativeProgress get() = latestProgress

    override fun start(): Promise<NativeCreationResult> = Promise.async {
        stateMachine.start()
        val startTime = System.currentTimeMillis()
        try {
            emitProgress(phase = "preparing")

            val destinationPath = when (request.destinationKind) {
                "file" -> request.destinationPath
                    ?: throw ArchiveDomainError.InvalidArgument("destinationPath required for file creation")
                else -> throw ArchiveDomainError.InvalidArgument("Unsupported destination kind: ${request.destinationKind}")
            }

            val entryInputs = request.entries.map { nativeEntry ->
                val data = nativeEntry.data?.let { arrayBuffer ->
                    arrayBuffer.toByteArray()
                }
                EntryInput(
                    kind = nativeEntry.kind,
                    sourcePath = nativeEntry.sourcePath,
                    archivePath = nativeEntry.archivePath,
                    data = data
                )
            }

            val plan = CreationPlan(
                entries = entryInputs,
                compressionProfile = request.compressionProfile,
                compressionLevel = request.compressionLevel?.toInt(),
                storeAlreadyCompressed = request.storeAlreadyCompressed ?: true,
                encryptionMethod = request.encryptionMethod,
                encryptionPassword = request.encryptionPassword
            )

            val output = LocalFileOutput(path = destinationPath)

            emitProgress(phase = "compressing")

            val useCase = CreateArchiveUseCase(engine)
            val result = useCase.execute(
                plan = plan,
                output = output,
                onProgress = { snapshot ->
                    emitProgress(
                        phase = "compressing",
                        processedBytes = snapshot.processedBytes,
                        totalBytes = snapshot.totalBytes,
                        processedEntries = snapshot.processedEntries.toDouble(),
                        totalEntries = snapshot.totalEntries?.toDouble(),
                        currentEntry = snapshot.currentEntry
                    )
                }
            )

            emitProgress(phase = "finalizing")

            val duration = System.currentTimeMillis() - startTime
            stateMachine.succeed()

            NativeCreationResult(
                operationId = operationId,
                entryCount = result.entryCount.toDouble(),
                inputBytes = result.inputBytes,
                outputBytes = result.outputBytes,
                durationMs = duration.toDouble(),
                atomicWriteApplied = false,
                warnings = emptyArray()
            )
        } catch (e: ArchiveDomainError) {
            stateMachine.fail()
            throw e
        } catch (e: Exception) {
            stateMachine.fail()
            throw e
        }
    }

    override fun cancel(): Boolean = stateMachine.cancel()

    override fun onProgress(callback: (progress: NativeProgress) -> Unit): () -> Unit {
        listeners.add(callback)
        return { listeners.remove(callback) }
    }

    private fun emitProgress(
        phase: String,
        processedBytes: Long = 0L,
        totalBytes: Long? = null,
        processedEntries: Double = 0.0,
        totalEntries: Double? = null,
        currentEntry: String? = null
    ) {
        val percentage = when {
            totalBytes != null && totalBytes > 0 ->
                minOf(processedBytes.toDouble() / totalBytes.toDouble() * 100.0, 100.0)
            totalEntries != null && totalEntries > 0 ->
                minOf(processedEntries / totalEntries * 100.0, 100.0)
            else -> null
        }
        val snapshot = NativeProgress(
            operationId = operationId,
            phase = phase,
            processedBytes = processedBytes,
            totalBytes = totalBytes,
            processedEntries = processedEntries,
            totalEntries = totalEntries,
            currentEntry = currentEntry,
            bytesPerSecond = null,
            estimatedSecondsRemaining = null,
            percentage = percentage
        )
        latestProgress = snapshot
        listeners.forEach { it(snapshot) }
    }
}
